lib/http_clients/creditas/token_server.ex

```diff
@@ -0,0 +1,112 @@
+defmodule HttpClients.Creditas.TokenServer do
+  @moduledoc """
+  An agent that stores a valid Creditas API token.
+
+  First setup it on your `application.ex`, e.g.:
+
+  ```elixir
+  config = [
+    url: "https://auth-staging.creditas.com.br/api/internal_clients/tokens",
+    credentials: %{
+      username: "someuser",
+      password: "password123",
+      grant_type: "password"
+    }
+  ]
+  children = [{HttpClients.Creditas.TokenServer, name: MyApp.CreditasTokenServer, config: config}]
+  Supervisor.start_link(children, name: MyApp.Supervisor)
+  ```
+
+  Then you can retrieve the token:
+
+  ```elixir
+  # get token and refresh it if is expired or at most 30 seconds before expiring
+  HttpClients.Creditas.TokenServer.get_token(MyApp.CreditasTokenServer)
+
+  # get token and refresh it if is expired or at most 120 seconds before expiring
+  HttpClients.Creditas.TokenServer.get_token(MyApp.CreditasTokenServer, 120)
+  ```
+  """
+
+  use Agent
+  require Logger
+
+  alias HttpClients.Creditas.Token
+
+  defguardp is_token_server(server) when is_pid(server) or is_atom(server)
+
+  @spec start_link(keyword()) :: {:ok, pid()} | {:error, any()} | no_return()
+  def start_link(opts) when is_list(opts) do
+    {config, opts} = Keyword.pop!(opts, :config)
+
+    Agent.start_link(
+      fn ->
+        {:ok, token} = request_new_token(config)
+        Logger.info("#{__MODULE__} started with pid #{inspect(self())}")
+        %{token: token, config: config}
+      end,
+      opts
+    )
+  end
+
+  @doc "Request an authenticated token to Creditas"
+  @spec request_new_token([{:credentials, map} | {:url, binary}, ...]) ::
+          {:error, any} | {:ok, Token.t()}
+  def request_new_token(url: url, credentials: credentials)
```

Comment:
```suggestion
  defp request_new_token(url: url, credentials: credentials)
```
essa n podia ser privada?

Reply:
poderia. eu curto pq é uma forma de pegar diretamente o token, mas sem problema esconder ela tmb
@bcredi/developers o que acham?

Reply:
se chamar ela diretamente, nosso Agent não vai ficar com o Token desatualizado? 🤔

Reply:
ela só faz a requisição pra API e builda o token :)

Reply:
saquei

Reply:
@willianschelles acho que essa função nao atualiza o agent, então o token fica intocado..
@youalreadydid Acho que entendi seu ponto, deixar essa função pra gente ter como uma "ferramenta" pra pegar o token se necessário, isso né?
Eu comentei mais pensando em alguem que venha de fora e começe a usar a lib, acho que ter uma única função pra pegar o token facilita o uso da lib.. 

Talvez renomear essa função pra `get_raw_token()` deixe mais claro que esse não é a função "principal" de pegar o token?

Reply:
entre
1 - `get_raw_token/1` + `get_token/1`
2 - `request_new_token/1` + `get_token/1`
eu prefiro o 2 pq o `raw_token` me remete ao payload da chamada ao invés da struct `%Token{}`.

@Gabee01 mas pensei que pode ser `get_new_token/1`, me parece que expressa mais o conceito

Reply:
mas se for `get_new_token/1` talvez deveria salvar esse token no state tmb

Reply:
vai travar o corsa piazada

Reply:
@Gabee01 @willianschelles mudei ali pra ter o `get_token` e o `get_new_token`, se puderem olhar agora